extensions/wikia/DefaultMessages/DefaultMessagesCache.php

```diff
@@ -0,0 +1,521 @@
+			$smallConds, __METHOD__ );
+
+		while ( $row = $dbr->fetchObject( $res ) ) {
+			$cache[$row->page_title] = ' ' . Revision::getRevisionText( $row );
+		}
+		$dbr->freeResult( $res );
+
+		$cache['VERSION'] = MSG_CACHE_VERSION;
+		wfProfileOut( __METHOD__ );
+		return $cache;
+	}
+
+	/**
+	 * Shortcut to update caches.
+	 *
+	 * @param $cache Array: cached messages with a version.
+	 * @param $cacheKey String: Identifier for the cache.
+	 * @param $memc Bool: Wether to update or not memcache.
+	 * @param $code String: Language code.
+	 * @return False on somekind of error.
+	 */
+	protected function saveToCaches( $cache, $memc = true, $code = false ) {
+		wfProfileIn( __METHOD__ );
+
+		if ( F::app()->wg->AllowMemcacheDisable && ( F::app()->wg->AllowMemcacheWrites == false ) ) {
+			return true;
+		}
+
+		$cacheKey = $this->memcKey( 'messages', $code );
+		$statusKey = $this->memcKey( 'messages', $code, 'status' );
+
+		$success = $this->mMemc->add( $statusKey, 'loading', MSG_LOAD_TIMEOUT );
+		if ( !$success ) {
+			wfProfileOut( __METHOD__ );
+			return true; # Other process should be updating them now
+		}
+
+		$i = 0;
+		if ( $memc ) {
+			# Save in memcached
+			# Keep trying if it fails, this is kind of important
+
+			for ($i=0; $i<20 &&
+				!$this->mMemc->set( $cacheKey, $cache, $this->mExpiry );
+				$i++ ) {
+				usleep(mt_rand(500000,1500000));
+			}
+		}
+
+		# Save to local cache
+		if ( $this->mLocalMessageCache !== false ) {
+			$serialized = serialize( $cache );
+			$hash = md5( $serialized );
+			$this->mMemc->set( $this->memcKey( 'messages', $code, 'hash' ), $hash, $this->mExpiry );
+			if ($this->mLocalMessageCacheSerialized) {
+				$this->saveToLocal( $serialized, $hash, $code );
+			} else {
+				$this->saveToScript( $cache, $hash, $code );
+			}
+		}
+
+		if ( $i == 20 ) {
+			$this->mMemc->set( $statusKey, 'error', 60*5 );
+			wfDebug( "MemCached set error in MessageCache: restart memcached server!\n" );
+			$success = false;
+		} else {
+			$this->mMemc->delete( $statusKey );
+			$success = true;
+		}
+		wfProfileOut( __METHOD__ );
+		return $success;
+	}
+
+	/**
+	 * Returns success
+	 * Represents a write lock on the messages key
+	 */
+	function lock($key) {
+		if ( !$this->mUseCache ) {
+			return true;
+		}
+
+		$lockKey = $key . ':lock';
+		for ($i=0; $i < MSG_WAIT_TIMEOUT && !$this->mMemc->add( $lockKey, 1, MSG_LOCK_TIMEOUT ); $i++ ) {
+			sleep(1);
+		}
+
+		return $i >= MSG_WAIT_TIMEOUT;
+	}
+
+	function unlock($key) {
+		if ( !$this->mUseCache ) {
+			return;
+		}
+
+		$lockKey = $key . ':lock';
+		$this->mMemc->delete( $lockKey );
+	}
+
+	/**
+	 * Get a message from the MediaWiki namespace, with caching. The key must
+	 * first be converted to two-part lang/msg form if necessary.
+	 *
+	 * @param $title String: Message cache key with initial uppercase letter.
+	 * @param $code String: code denoting the language to try.
+	 */
+	function get( $title, $code, $useDB=true ) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Reduce the number of returns of this function 5, down to the maximum allowed 3. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS1142)